caddytls/ctmonitor.go

```diff
@@ -0,0 +1,414 @@
+}
+
+// CertSpotterResponse is the json response formatting that is used in the program.
+type CertSpotterResponse struct {
+	ID           string   `json:"id"`
+	TBSSHA256    string   `json:"tbs_sha256"`
+	DNSNames     []string `json:"dns_names"`
+	PubKeySha256 string   `json:"pubkey_sha256"`
+	Issuer       struct {
+		Name         string `json:"name"`
+		PubKeySha256 string `json:"pubkey_sha256"`
+	} `json:"issuer"`
+	NotBefore string `json:"not_before"`
+	NotAfter  string `json:"not_after"`
+	Cert      struct {
+		ID     string `json:"id"`
+		Type   string `json:"type"`
+		SHA256 string `json:"sha256"`
+		Data   string `json:"data"`
+	} `json:"cert"`
+}
+
+// QueryConfig is the configuration information for querying.
+type QueryConfig struct {
+	CTConfig
+	Query string
+	Index int
+}
+
+// CTMonitor holds a lot of the data that is used throughout the plugin.
+type CTMonitor struct {
+	// caddyCerts consists of the certs that caddy serves. Key = bytes of each cert, cast to a string
+	// caddyCertsSANs are the Subject Alternate Names that Caddy is hosting.
+	// used for individual queries, is an element in caddyCertsSANS
+	caddyCerts     map[string]struct{}
+	caddyCertsSANs []string
+	domainName     string
+
+	// issuance stored in the map entry in case the keys don't match.
+	// CertSpotterResponses, used in comparisons against the caddyCerts. map of bytes to issuance.
+	issuance       CertSpotterResponse
+	retrievedCerts map[string]CertSpotterResponse
+
+	// configuration struct for the CTMonitor
+	config QueryConfig
+
+	// used for terminating the routine when the shutdown event is called.
+	done chan struct{}
+
+	// retryAfter will contain the string arg from the response, used for next iteration.
+	// timeToWait is the numerical value of retryAfter.
+	retryAfter string
+	timeToWait int
+
+	// these variables are used while processing a CertSpotter response.
+	// biggestID is the most recent certID returned from CertSpotter.
+	// issuanceID is compared with biggestID to determine the most recent cert to store.
+	// lastIssuance tells it when to compare the biggestID with the issuanceID.
+	biggestID    int
+	issuanceID   int
+	lastIssuance bool
+}
+
+// BuildMapEntry adds only certificates to the map of retrievedCerts, and it gets the issuanceID and compares it to the biggestID we have seen so far.
+func buildMapEntry() error {
+	bytes, err := base64.StdEncoding.DecodeString(ctmonitor.issuance.Cert.Data)
+	if err != nil {
+		return fmt.Errorf("Decoding failed: %v", err)
+	}
+	aKey := string(bytes)
+	if _, ok := ctmonitor.retrievedCerts[aKey]; ok {
+		return nil
+	}
+	if ctmonitor.issuance.Cert.Type == "precert" {
+		return nil
+	}
+
+	ctmonitor.retrievedCerts[aKey] = ctmonitor.issuance
+	if ctmonitor.lastIssuance {
+		ctmonitor.issuanceID, err = strconv.Atoi(ctmonitor.issuance.ID)
+		if err != nil {
+			return fmt.Errorf("failed to convert issuance.ID to an int: %v", err)
+		}
+		if ctmonitor.issuanceID > ctmonitor.biggestID {
```

Comment:
Since issuance IDs are opaque, you can't make any inferences about the relation of two different IDs.  I'm curious what the motivation behind this comparison is?